pkg/metallb/frr.go

```diff
@@ -0,0 +1,589 @@
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].KeepaliveTime = &keepAlive
+
+	return builder
+}
+
+// WithConnectTime defines the time to wait until trying to connect to a BGP peer.
+func (builder *FrrConfigurationBuilder) WithConnectTime(connectTime metav1.Duration, routerIndex,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Defines a connect retry hold time for a specific neighbor in FrrConfiguration %s with holdTime %s ",
+		builder.Definition.Name, connectTime)
+
+	duration := connectTime.Duration
+
+	if duration < time.Second || duration > 65535*time.Second {
+		glog.V(100).Infof("A valid connect time is between 1-65535")
+
+		builder.errorMsg = "frrConfiguration 'connectTime' value is not valid"
+	}
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].ConnectTime = &connectTime
+
+	return builder
+}
+
+// WithPort defines the tcp port used to make BGP adjacency.
+func (builder *FrrConfigurationBuilder) WithPort(port uint16, routerIndex,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Defines the tcp port to establish adjaceny to a neighbor in FrrConfiguration %s port number %d",
+		builder.Definition.Name, port)
+
+	if port > 16384 {
+		glog.V(100).Infof("Invalid port number can not be greater then 16384: %d", port)
+
+		builder.errorMsg = fmt.Sprintf("invalid port number: %d", port)
+	}
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
```

Comment:
This can be removed if we short-circuit return builder.

Reply:
removed